### ISSUE TITLE
Deal with empty (and `$Null`) inputs

### ISSUE THREAD

**Issue by iRon7** [iRon7/Join-Object#19]
Considered how to deal with `$Null` and empty array inputs:
- `@() |Join $Object` 
  This situation could potentially occur from a chained (inner) join.
  But at the other hand, could useful as initial "object" in an loop: `... |% { $a = $Null } { $a = $a |Join $_ }`
- `$Object |Join $Null` or `$Object |Join @()`

Current situation:
```PowerShell
PS C:\> $Null |Join $Employee |Ft

Id     Name               Country            Department                 Age      ReportsTo
--     ----               -------            ----------                 ---      ---------
{1, 1} {Aerts, Aerts}     {Belgium, Belgium} {Sales, Sales}             {40, 40} {5, 5}
{2, 2} {Bauer, Bauer}     {Germany, Germany} {Engineering, Engineering} {31, 31} {4, 4}
{3, 3} {Cook, Cook}       {England, England} {Sales, Sales}             {69, 69} {1, 1}
{4, 4} {Duval, Duval}     {France, France}   {Engineering, Engineering} {21, 21} {5, 5}
{5, 5} {Evans, Evans}     {England, England} {Marketing, Marketing}     {35, 35} {, }
{6, 6} {Fischer, Fischer} {Germany, Germany} {Engineering, Engineering} {29, 29} {4, 4}

PS C:\> @() |Join $Employee |Ft

Id     Name               Country            Department                 Age      ReportsTo
--     ----               -------            ----------                 ---      ---------
{1, 1} {Aerts, Aerts}     {Belgium, Belgium} {Sales, Sales}             {40, 40} {5, 5}
{2, 2} {Bauer, Bauer}     {Germany, Germany} {Engineering, Engineering} {31, 31} {4, 4}
{3, 3} {Cook, Cook}       {England, England} {Sales, Sales}             {69, 69} {1, 1}
{4, 4} {Duval, Duval}     {France, France}   {Engineering, Engineering} {21, 21} {5, 5}
{5, 5} {Evans, Evans}     {England, England} {Marketing, Marketing}     {35, 35} {, }
{6, 6} {Fischer, Fischer} {Germany, Germany} {Engineering, Engineering} {29, 29} {4, 4}

PS C:\> $Employee |Join $Null |Ft
InvalidOperation: C:\Join.ps1:345
Line |
 345 |  … in $Keys) { $Properties.Add($Key, $Object.psobject.properties[$Key].V …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot index into a null array.


Id Name  Country Department Age ReportsTo
-- ----  ------- ---------- --- ---------
 1 Aerts Belgium Sales       40         5

PS C:\> $Employee |Join @() |Ft
InvalidOperation: C:\Join.ps1:454
Line |
 454 |                  ([ref]$RightKeys).Value = $Right.get_Keys()
     |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | You cannot call a method on a null-valued expression.

MethodInvocationException: C:\Join.ps1:371
Line |
 371 |                      if (!$Expressions.Contains($Key)) {
     |                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Exception calling "Contains" with "1" argument(s): "Value cannot be null. (Parameter 'key')"

**Comment by iRon7** [comment]
This doesn't return anything because there is no relation between the `id`'s:
```PowerShell
@{id = 1; name = 'one'} |Join @{id = 2; name = 'Two'} -On id
```

### Expected:
Therefore it is expected that chaining it to a third object should still return nothing:
```PowerShell
@{id = 1; name = 'one'} |Join @{id = 2; name = 'Two'} -On id |Join @{id = 3; name = 'Three'} -On id
```

### Actual
But instead is does a **self-join** on the last object and returns:
 ```
name           id
----           --
{Three, Three}  3
```

**Comment by iRon7** [comment]
Fixed in version `3.4.6`

```PowerShell
@() |Join $Employee | Should -benull

$Actual = $Employee |Join $Null 
$Expected = [PSCustomObject]@{Id = 1; Name = 'Aerts'; Country = 'Belgium'; Department = 'Sales'; Age = 40; ReportsTo = 5}
Compare-PSObject $Actual $Expected | Should -BeNull

$Employee |Join @() | Should -benull

@{id = 1; name = 'one'} |Join @{id = 2; name = 'Two'} -On id | Should -benull

@{id = 1; name = 'one'} |Join @{id = 2; name = 'Two'} -On id |Join @{id = 3; name = 'Three'} -On id | Should -benull

@() |Join @{id = 3; name = 'Three'} -On id | Should -benull

$Actual = Join @{id = 3; name = 'Three'} -On id
$Expected = [PSCustomObject]@{name = [array]('Three', 'Three'); id = 3}
Compare-PSObject $Actual $Expected | Should -BeNull
```

**Comment by iRon7** [comment]
Version 3.8 (prerelease), joining with `$Null` doesn't return anything

```PowerShell
$Employee |Join $Null | Should -benull
```
